packages/flutter_tools/test/integration.shard/overall_experience_test.dart

```diff
@@ -0,0 +1,568 @@
+final String flutterBin = fileSystem.path.join(flutterRoot, 'bin', 'flutter');
+
+typedef LineHandler = String/*?*/ Function(String line);
+
+abstract class Transition {
+  const Transition({this.handler, this.logging});
+  final LineHandler/*?*/ handler;
+  final bool/*?*/ logging;
+
+  bool matches(String line);
+
+  @protected
+  bool lineMatchesPattern(String line, Pattern pattern) {
+    if (pattern is String) {
+      return line == pattern;
+    }
+    return line.contains(pattern);
+  }
+
+  @protected
+  String describe(Pattern pattern) {
+    if (pattern is String) {
+      return '"$pattern"';
+    }
+    if (pattern is RegExp) {
+      return '/${pattern.pattern}/';
+    }
+    return '$pattern';
+  }
+}
+
+class Barrier extends Transition {
+  const Barrier(this.pattern, {LineHandler/*?*/ handler, bool/*?*/ logging}) : super(handler: handler, logging: logging);
+  final Pattern pattern;
+
+  @override
+  bool matches(String line) => lineMatchesPattern(line, pattern);
+
+  @override
+  String toString() => describe(pattern);
+}
+
+class Multiple extends Transition {
+  Multiple(List<Pattern> patterns, {
+    LineHandler/*?*/ handler,
+    bool/*?*/ logging,
+  }) : _originalPatterns = patterns,
+       patterns = patterns.toList(),
+       super(handler: handler, logging: logging);
+
+  final List<Pattern> _originalPatterns;
+  final List<Pattern> patterns;
+
+  @override
+  bool matches(String line) {
+    for (int index = 0; index < patterns.length; index += 1) {
+      if (lineMatchesPattern(line, patterns[index])) {
+        patterns.removeAt(index);
+        break;
+      }
+    }
+    return patterns.isEmpty;
+  }
+
+  @override
+  String toString() {
+    return _originalPatterns.map(describe).join(', ') + ' (matched ${_originalPatterns.length - patterns.length} so far)';
+  }
+}
+
+class ProcessTestResult {
+  const ProcessTestResult(this.exitCode, this.stdout, this.stderr);
+  final int exitCode;
+  final List<String> stdout;
+  final List<String> stderr;
+
+  @override
+  String toString() => 'exit code $exitCode\nstdout:\n  ${stdout.join('\n  ')}\nstderr:\n  ${stderr.join('\n  ')}\n';
+}
+
+String clarify(String line) {
+  return line.runes.map<String>((int rune) {
+    if (rune >= 0x20 && rune <= 0x7F) {
+      return String.fromCharCode(rune);
+    }
+    switch (rune) {
+      case 0x00: return '<NUL>';
+      case 0x07: return '<BEL>';
+      case 0x08: return '<TAB>';
+      case 0x09: return '<BS>';
+      case 0x0A: return '<LF>';
+      case 0x0D: return '<CR>';
+    }
+    return '<${rune.toRadixString(16).padLeft(rune <= 0xFF ? 2 : rune <= 0xFFFF ? 4 : 5, '0')}>';
+  }).join('');
+}
+
+void printClearly(String line) {
+  print(clarify(line));
+}
+
+Future<ProcessTestResult> runFlutter(
+  List<String> arguments,
+  String workingDirectory,
+  List<Transition> transitions, {
+  bool debug = false,
+  bool logging = true,
+}) async {
+  final Process process = await processManager.start(
```

Comment:
I think it is possible for the test to fail and the next one to start before the process gets killed. I noticed that if writing the PID file fails the expect on L279, then the next test will always fail (on windows) because the previous flutter instance is still going. You might need something like a `finally` block that kills the process if a test exception is thrown

Reply:
There's not much I can do about this because the subprocess is a shell process, not the Dart process, so killing it doesn't actually do anything useful.

I've made the first test avoid expects in the callbacks, which addresses the immediate concern you mention. Not much I can do to enforce that though.

That said, I actually don't think you're right. Each test is generating its own temporary directory with its own PID file, so zombie flutter processes from earlier tests shouldn't affect later tests. (Unless you're saying on Windows you can never run `flutter` twice at the same time? In which case, I wasn't aware of that. Still not much I can do though.)

Reply:
I think it's specific to how we `flutter run` on the tester device - we copy the kernel artifacts to a generic location under build/ , but instead it should probably be to a temp directory. Filled https://github.com/flutter/flutter/issues/75534

But at any rate, after the test failed and exited I still have some zombie flutter_tester processes running on windows. Maybe write a `q` into stdin to try and shut them down gracefully?

Reply:
Added this:
```patch
-  final int exitCode = await process.exitCode;
+  final int exitCode = await process.exitCode
+    ..timeout(const Duration(seconds: 15), () {
+      print('(process is not quitting, trying to send a "q" just in case that helps)');
+      print('(a functional test should never reach this point)');
+      process.stdin.write('q');
+    });
```

Reply:
...I swear I tested that, but somehow the code above (and the code I actually landed) was bogus. Anyway, I've fixed that now...